scanpy/plotting/_tools/__init__.py

```diff
@@ -721,52 +726,92 @@ def embedding_density(
                          'Please specify a group from this covariate to plot.')
 
-    if (group is not None) and (group not in adata.obs[groupby].cat.categories):
-        raise ValueError('Please specify a group from the `.obs` category over which the density '
-                         'was calculated.')
-
     if (np.min(adata.obs[key]) < 0) or (np.max(adata.obs[key]) > 1):
         raise ValueError('Densities should be scaled between 0 and 1.')
 
+    if wspace is None:
+        #  try to set a wspace that is not too large or too small given the
+        #  current figure size
+        wspace = 0.75 / rcParams['figure.figsize'][0] + 0.02
+
     # Define plotting data
     dens_values = -np.ones(adata.n_obs)
     dot_sizes = np.ones(adata.n_obs)*bg_dotsize
 
-    if group is not None:
-        group_mask = (adata.obs[groupby] == group)
-        dens_values[group_mask] = adata.obs[key][group_mask]
-        dot_sizes[group_mask] = np.ones(sum(group_mask))*fg_dotsize
-
-    else:
-        dens_values = adata.obs[key]
-        dot_sizes = np.ones(adata.n_obs)*fg_dotsize
-
     # Make the color map
     if isinstance(color_map, str):
         cmap = cm.get_cmap(color_map)
     else:
         cmap = color_map
 
-    #norm = colors.Normalize(vmin=-1, vmax=1)
-    adata_vis = adata.copy()
-    adata_vis.obs['Density'] = dens_values
-
     norm = colors.Normalize(vmin=vmin, vmax=vmax)
     cmap.set_over('black')
     cmap.set_under('lightgray')
 
-    # Ensure title is blank as default
-    if 'title' not in kwargs:
-        title=""
+    if group is not None and isinstance(group, abc.Sequence):
+        from matplotlib import gridspec
+        # set up the figure
+        num_panels = len(group)
+        n_panels_x = min(ncols, num_panels)
+        n_panels_y = np.ceil(num_panels / n_panels_x).astype(int)
+        # each panel will have the size of rcParams['figure.figsize']
+        fig = pl.figure(figsize=(n_panels_x * rcParams['figure.figsize'][0] * (1 + wspace),
+                                 n_panels_y * rcParams['figure.figsize'][1]))
+        left = 0.2 / n_panels_x
+        bottom = 0.13 / n_panels_y
+        gs = gridspec.GridSpec(
+            nrows=n_panels_y, ncols=n_panels_x,
+            left=left, right=1 - (n_panels_x - 1) * left - 0.01 / n_panels_x,
+            bottom=bottom, top=1 - (n_panels_y - 1) * bottom - 0.1 / n_panels_y,
+            hspace=hspace, wspace=wspace,
+        )
+
+        axs = []
+        for count, group_name in enumerate(group):
+            if group_name not in adata.obs[groupby].cat.categories:
+                raise ValueError('Please specify a group from the `.obs` category over which the density '
+                                 'was calculated. Invalid group name: {}'.format(group_name))
+
+            ax = pl.subplot(gs[count])
+            group_mask = (adata.obs[groupby] == group_name)
+            adata.obs['_density'] = adata.obs[key][group_mask]
```

Comment:
Can we assume that we will not overwrite a `'_density'` slot in the original `AnnData` object? I created a new one for that reason. Not sure if `_VarName` can also be assumed to be reserved for `adata.obs` slots.

Reply:
good point. I didn't want to copy the whole AnnData object to save memory but indeed this could rewrite a valid obs. name. I will set up some random name to avoid this.